src/Avalonia.Base/Input/GestureRecognizers/VelocityTracker.cs

```diff
@@ -0,0 +1,402 @@
+                    r[j, i] = i < j ? 0.0 : Multiply(q.GetRow(j), a.GetRow(i));
+                }
+            }
+
+            // Solve R B = Qt W Y to find B. This is easy because R is upper triangular.
+            // We just work from bottom-right to top-left calculating B's coefficients.
+            // "m" isn't expected to be bigger than HistorySize=20, so allocation on stack is safe.
+            Span<double> wy = stackalloc double[m];  
+            for (int h = 0; h < m; h += 1)
+            {
+                wy[h] = y[h] * w[h];
+            }
+            for (int i = n - 1; i >= 0; i -= 1)
+            {
+                result.Coefficients[i] = Multiply(q.GetRow(i), wy);
+                for (int j = n - 1; j > i; j -= 1)
+                {
+                    result.Coefficients[i] -= r[i, j] * result.Coefficients[j];
+                }
+                result.Coefficients[i] /= r[i, i];
+            }
+
+            // Calculate the coefficient of determination (confidence) as:
+            //   1 - (sumSquaredError / sumSquaredTotal)
+            // ...where sumSquaredError is the residual sum of squares (variance of the
+            // error), and sumSquaredTotal is the total sum of squares (variance of the
+            // data) where each has been weighted.
+            double yMean = 0.0;
+            for (int h = 0; h < m; h += 1)
+            {
+                yMean += y[h];
+            }
+            yMean /= m;
+
+            double sumSquaredError = 0.0;
+            double sumSquaredTotal = 0.0;
+            for (int h = 0; h < m; h += 1)
+            {
+                double term = 1.0;
+                double err = y[h] - result.Coefficients[0];
+                for (int i = 1; i < n; i += 1)
+                {
+                    term *= x[h];
+                    err -= term * result.Coefficients[i];
+                }
+                sumSquaredError += w[h] * w[h] * err * err;
+                double v = y[h] - yMean;
+                sumSquaredTotal += w[h] * w[h] * v * v;
+            }
+
+            result.Confidence = sumSquaredTotal <= PrecisionErrorTolerance ? 1.0 :
+                                  1.0 - (sumSquaredError / sumSquaredTotal);
+
+            return result;
+        }
+
+        private static double Multiply(Span<double> v1, Span<double> v2)
+        {
+            double result = 0.0;
+            for (int i = 0; i < v1.Length; i += 1)
+            {
+                result += v1[i] * v2[i];
+            }
+            return result;
+        }
+        
+        private static double Norm(Span<double> v)
+        {
+            return Math.Sqrt(Multiply(v, v));
+        }
+
+        private readonly struct _Matrix
+        {
+            private readonly int _columns;
+            private readonly double[] _elements;
+
+            internal _Matrix(int rows, int cols)
+            {
+                _columns = cols;
+                _elements = new double[rows * cols];
```

Comment:
.NET ArrayPool technically can be used here. But we probably don't have a need to avoid every single allocation here.

Reply:
`_elements` needs to be initialized to all 0.0 I think for the algorithm to function -- ArrayPool<double> rents out an array in its  'Return' state right? I'll change it so _Matrix gets a Span<double> in its constructor -- and stackalloc it in the caller, as its only used 3 times.